rust/src/fvm/machine.rs

```diff
@@ -0,0 +1,264 @@
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::mem;
+use std::ptr;
+use std::sync::{atomic::AtomicU64, Mutex};
+
+use anyhow::Error;
+use blockstore::cgo::CgoBlockstore;
+use blockstore::{Block, Blockstore, MemoryBlockstore};
+use cid::Cid;
+use drop_struct_macro_derive::DropStructMacro;
+use ffi_toolkit::{
+    c_str_to_pbuf, catch_panic_response, raw_ptr, rust_str_to_c_str, FCPResponseStatus,
+};
+use fvm::externs::cgo::CgoExterns;
+use fvm::externs::Externs;
+use fvm::machine::{ApplyKind, ApplyRet, Machine};
+use fvm::message::Message;
+use fvm::Config;
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::encoding::RawBytes;
+use fvm_shared::version::NetworkVersion;
+use fvm_shared::MethodNum;
+use log::{error, info};
+use num_traits::FromPrimitive;
+use once_cell::sync::Lazy;
+
+use super::types::*;
+use crate::util::api::init_log;
+
+type CgoMachine = Machine<CgoBlockstore, CgoExterns>;
+
+static FVM_MAP: Lazy<Mutex<HashMap<u64, CgoMachine>>> =
+    Lazy::new(|| Mutex::new(HashMap::with_capacity(1)));
+
+const NEXT_ID: AtomicU64 = AtomicU64::new(0);
+
+fn add_fvm_machine(machine: CgoMachine) -> u64 {
+    let next_id = NEXT_ID.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+    let mut machines = FVM_MAP.lock().unwrap();
+    machines.insert(next_id, machine);
+    next_id
+}
+
+fn get_default_config() -> fvm::Config {
+    Config {
+        initial_pages: 1024, //FIXME
+        max_pages: 32768,    // FIXME
+        engine: wasmtime::Config::new(),
+    }
+}
+
+/// Note: the incoming args as u64 and odd conversions to i32/i64
+/// for some types is due to the generated bindings not liking the
+/// 32bit types as incoming args
+///
+#[no_mangle]
+#[cfg(not(target_os = "windows"))]
+pub unsafe extern "C" fn fil_create_fvm_machine(
+    fvm_version: fil_FvmRegisteredVersion,
+    chain_epoch: u64,
+    token_amount: u64,
+    network_version: u64,
+    state_root_ptr: *const u8,
+    state_root_len: libc::size_t,
+    blockstore_id: u64,
+    externs_id: u64,
+) -> *mut fil_CreateFvmMachineResponse {
+    catch_panic_response(|| {
+        init_log();
+
+        info!("fil_create_fvm_machine: start");
+
+        let mut response = fil_CreateFvmMachineResponse::default();
+
+        let config = get_default_config();
+        let chain_epoch = chain_epoch as ChainEpoch;
+        let token_amount = TokenAmount::from_u64(token_amount);
```

Comment:
Can't be u64, needs to be u128 if safe for FFI, or a tuple of (hi, lo) u64s.

Reply:
We usually just use bigints.